File: 034_LeafSimilarTrees.cpp
// Topic: Binary Tree / DFS
// Leaf-Similar Trees: Check if two binary trees have the same sequence of leaf node values.
// Time Complexity: O(n+m)
// Space Complexity: O(h1+h2)

#include <bits/stdc++.h>
using namespace std;
struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

void leaves(TreeNode* root, vector<int>& out)
{
    if (!root)
        return;
    if (!root->left && !root->right)
    {
        out.push_back(root->val);
        return;
    }
    leaves(root->left, out);
    leaves(root->right, out);
}

bool leafSimilar(TreeNode* r1, TreeNode* r2)
{
    vector<int> a, b;
    leaves(r1, a);
    leaves(r2, b);
    return a == b;
}

int main()
{
    TreeNode* a = new TreeNode(3);
    a->left = new TreeNode(5);
    a->right = new TreeNode(1);
    a->left->left = new TreeNode(6);
    a->left->right = new TreeNode(2);
    a->left->right->left = new TreeNode(7);
    a->left->right->right = new TreeNode(4);
    TreeNode* b = new TreeNode(3);
    b->left = new TreeNode(5);
    b->right = new TreeNode(1);
    b->left->left = new TreeNode(6);
    b->left->right = new TreeNode(7);
    b->right->left = new TreeNode(4);
    b->right->right = new TreeNode(2);
    cout << (leafSimilar(a, b) ? "true" : "false") << "\n";  // true
    TreeNode* c = new TreeNode(1);
    c->left = new TreeNode(2);
    c->right = new TreeNode(3);
    cout << (leafSimilar(a, c) ? "true" : "false") << "\n";  // false
    return 0;
}

File: 034_LeafSimilarTrees.py
"""
Topic: Tree / DFS
Leaf-Similar Trees: compare leaf traversal sequences
Time Complexity: O(n + m)
Space Complexity: O(h + k)
"""

from typing import Optional, List

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def leaf_sequence(root: Optional[TreeNode], out: List[int]):
    if not root: return
    if not root.left and not root.right:
        out.append(root.val); return
    leaf_sequence(root.left, out)
    leaf_sequence(root.right, out)

def leaf_similar(root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
    a, b = [], []
    leaf_sequence(root1, a)
    leaf_sequence(root2, b)
    return a == b

if __name__ == '__main__':
    # tree1 leaves: [6,7,4,2]
    t1 = TreeNode(3, TreeNode(5, TreeNode(6), TreeNode(2, TreeNode(7), TreeNode(4))), TreeNode(1))
    t2 = TreeNode(3, TreeNode(5, TreeNode(6), TreeNode(2, TreeNode(7), TreeNode(4))), TreeNode(1))
    print(leaf_similar(t1, t2))  # True
    print(leaf_similar(t1, TreeNode(1)))  # False
